process/sync/baseSync.go

```diff
@@ -316,104 +313,72 @@ func (boot *baseBootstrap) cleanCachesAndStorageOnRollback(header data.HeaderHan
-	store dataRetriever.StorageService,
-	blackListHandler process.BlackListHandler,
-	watcher process.NetworkConnectionWatcher,
-	requestedItemsHandler dataRetriever.RequestedItemsHandler,
-) error {
-	if check.IfNil(blkc) {
+func checkBootstrapNilParameters(arguments ArgBaseBootstrapper) error {
+	if check.IfNil(arguments.ChainHandler) {
 		return process.ErrNilBlockChain
 	}
-	if check.IfNil(rounder) {
+	if check.IfNil(arguments.Rounder) {
 		return process.ErrNilRounder
 	}
-	if check.IfNil(blkExecutor) {
-		return process.ErrNilBlockExecutor
+	if check.IfNil(arguments.BlockProcessor) {
+		return process.ErrNilBlockProcessor
 	}
-	if check.IfNil(hasher) {
+	if check.IfNil(arguments.Hasher) {
 		return process.ErrNilHasher
 	}
-	if check.IfNil(marshalizer) {
+	if check.IfNil(arguments.Marshalizer) {
 		return process.ErrNilMarshalizer
 	}
-	if check.IfNil(forkDetector) {
+	if check.IfNil(arguments.ForkDetector) {
 		return process.ErrNilForkDetector
 	}
-	if check.IfNil(resolversFinder) {
-		return process.ErrNilResolverContainer
+	if check.IfNil(arguments.RequestHandler) {
+		return process.ErrNilRequestHandler
 	}
-	if check.IfNil(shardCoordinator) {
+	if check.IfNil(arguments.ShardCoordinator) {
 		return process.ErrNilShardCoordinator
 	}
-	if check.IfNil(accounts) {
+	if check.IfNil(arguments.Accounts) {
 		return process.ErrNilAccountsAdapter
 	}
-	if check.IfNil(store) {
+	if check.IfNil(arguments.Store) {
 		return process.ErrNilStore
 	}
-	if check.IfNil(blackListHandler) {
+	if check.IfNil(arguments.BlackListHandler) {
 		return process.ErrNilBlackListHandler
 	}
-	if check.IfNil(watcher) {
+	if check.IfNil(arguments.NetworkWatcher) {
 		return process.ErrNilNetworkWatcher
 	}
-	if check.IfNil(requestedItemsHandler) {
-		return dataRetriever.ErrNilRequestedItemsHandler
+	if check.IfNil(arguments.MiniBlocksResolver) {
+		return process.ErrNilMiniBlocksResolver
 	}
-
 	return nil
 }
 
 func (boot *baseBootstrap) requestHeadersFromNonceIfMissing(
 	nonce uint64,
-	haveHeaderInPoolWithNonce func(uint64) bool,
-	hdrRes dataRetriever.HeaderResolver) {
-
-	boot.requestedItemsHandler.Sweep()
-
+	haveHeaderInPoolWithNonce func(uint64) bool) {
 	nbRequestedHdrs := 0
 	maxNonce := core.MinUint64(nonce+process.MaxHeadersToRequestInAdvance-1, boot.forkDetector.ProbableHighestNonce())
 	for currentNonce := nonce; currentNonce <= maxNonce; currentNonce++ {
-		key := fmt.Sprintf("%d-%d", boot.shardCoordinator.SelfId(), currentNonce)
-		if boot.requestedItemsHandler.Has(key) {
-			continue
-		}
-
 		haveHeader := haveHeaderInPoolWithNonce(nonce)
 		if !haveHeader {
-			err := hdrRes.RequestDataFromNonce(currentNonce, boot.epochHandler.Epoch())
-			if err != nil {
-				log.Debug("RequestDataFromNonce", "error", err.Error())
-				continue
-			}
-
-			err = boot.requestedItemsHandler.Add(key)
-			if err != nil {
-				log.Trace("add requested item with error", "error", err.Error())
+			if boot.shardCoordinator.SelfId() == sharding.MetachainShardId {
```

Comment:
maybe you can change logic here to reduce complexity if haveHeader { continue}...

Reply:
and also add  one more method in _blockBootstrapper_ interface to can requestHeaderByNonce(nonde uint64) and also if instruction where is checked shard id can be removed.

Reply:
Done